File: packages/python-sdk/src/mogbank/wallet.py
"""Wallet module — manage agent wallets, balances, and deposit addresses."""

from __future__ import annotations
from typing import Optional, List
from dataclasses import dataclass

from mogbank.client import MogBankClient
from mogbank.types import WalletInfo, WalletStatus, ApiResponse, Currency


@dataclass
class CreateWalletParams:
    agent_id: str
    currency: Currency = Currency.USDC


@dataclass
class WalletResponse(WalletInfo):
    pass


class WalletModule:
    def __init__(self, client: MogBankClient):
        self.client = client

    def create(self, params: CreateWalletParams) -> WalletInfo:
        result = self.client.request(
            "POST",
            "/api/v1/wallets",
            {
                "agent_id": params.agent_id,
                "currency": params.currency.value,
            },
            idempotency_key=f"wallet_create_{params.agent_id}_{hash(str(params))}",
            sign_with_ed25519=True,
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Wallet creation failed")
        return WalletInfo(**result.data)

    def list(self, agent_id: str) -> List[WalletInfo]:
        result = self.client.request(
            "GET", f"/api/v1/wallets?agent_id={agent_id}"
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Failed to list wallets")
        return [WalletInfo(**w) for w in result.data]

    def get(self, wallet_id: str) -> WalletInfo:
        result = self.client.request("GET", f"/api/v1/wallets/{wallet_id}")
        if not result.success or not result.data:
            raise Exception(result.error or "Wallet not found")
        return WalletInfo(**result.data)

    def balance(self, wallet_id: str) -> dict:
        wallet = self.get(wallet_id)
        return {"balance": wallet.balance, "currency": wallet.currency.value}

    def list_all(self) -> List[WalletInfo]:
        result = self.client.request("GET", "/api/v1/admin/wallets")
        if not result.success or not result.data:
            raise Exception(result.error or "Failed to list wallets")
        return [WalletInfo(**w) for w in result.data]

    def set_status(self, wallet_id: str, status: WalletStatus) -> WalletInfo:
        result = self.client.request(
            "POST",
            f"/api/v1/wallets/{wallet_id}/status",
            {"status": status.value},
            idempotency_key=f"wallet_sts_{wallet_id}_{status.value}",
            sign_with_ed25519=True,
        )
        if not result.success or not result.data:
            raise Exception(result.error or "Status update failed")
        return WalletInfo(**result.data)

File: packages/typescript-sdk/src/wallet.ts
/**
 * Wallet module — manage agent wallets, balances, and deposit addresses.
 */
import { MogBankClient } from './client';
import { WalletInfo, ApiResponse } from './types';

export type WalletResponse = WalletInfo;

export interface CreateWalletParams {
  agent_id: string;
  currency?: 'USDC' | 'UNIT';
}

export class WalletModule {
  constructor(private client: MogBankClient) {}

  /**
   * Create a new wallet for an agent.
   */
  async create(params: CreateWalletParams): Promise<WalletInfo> {
    const result = await this.client.request<ApiResponse<WalletInfo>>(
      'POST',
      '/api/v1/wallets',
      params,
      {
        idempotencyKey: `wallet_create_${params.agent_id}_${Date.now()}`,
        signWithEd25519: true,
      }
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Wallet creation failed');
    }
    return result.data;
  }

  /**
   * Get all wallets for the authenticated agent.
   */
  async list(agentId: string): Promise<WalletInfo[]> {
    const result = await this.client.request<ApiResponse<WalletInfo[]>>(
      'GET',
      `/api/v1/wallets?agent_id=${encodeURIComponent(agentId)}`
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Failed to list wallets');
    }
    return result.data;
  }

  /**
   * Get a specific wallet by ID.
   */
  async get(walletId: string): Promise<WalletInfo> {
    const result = await this.client.request<ApiResponse<WalletInfo>>(
      'GET',
      `/api/v1/wallets/${walletId}`
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Wallet not found');
    }
    return result.data;
  }

  /**
   * Get wallet balance.
   */
  async balance(walletId: string): Promise<{ balance: number; currency: string }> {
    const wallet = await this.get(walletId);
    return { balance: wallet.balance, currency: wallet.currency };
  }

  /**
   * Get all wallets across all agents (admin only).
   */
  async listAll(): Promise<WalletInfo[]> {
    const result = await this.client.request<ApiResponse<WalletInfo[]>>(
      'GET',
      '/api/v1/admin/wallets'
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Failed to list wallets');
    }
    return result.data;
  }

  /**
   * Freeze or unfreeze a wallet (admin only).
   */
  async setStatus(walletId: string, status: 'active' | 'frozen' | 'closed'): Promise<WalletInfo> {
    const result = await this.client.request<ApiResponse<WalletInfo>>(
      'POST',
      `/api/v1/wallets/${walletId}/status`,
      { status },
      {
        idempotencyKey: `wallet_sts_${walletId}_${status}_${Date.now()}`,
        signWithEd25519: true,
      }
    );
    if (!result.success || !result.data) {
      throw new Error(result.error || 'Status update failed');
    }
    return result.data;
  }
}